### PR TITLE
TICKET 8093: Google OAuth 2.0

### PR DESCRIPTION
**Pull request by Kanaduchi**
1. We can integrate any oauth system. Google is an example

2. If user does not exist in DB and login through oauth - he will be saved to db with special auth type. So for secure reasons he can't logon via regular email/password into Testlink. As we can't get this password from oauth

3. If user already exists in DB then he can logon via password or via oauth

### CODE REVIEW

**Comment by fmancardi** [comment]
Please do not use ONE PULL REQUEST For more than ONE feature and/or issue.
Please do not mix a FEATURE with a BUG FIX on same Pull request

**Comment by Kanaduchi** [comment]
Updated

File: gui/templates/login-model-marcobiedermann.tpl [hunk 1]
```diff
@@ -63,6 +63,14 @@
               <input type="submit" value="{$labels.btn_login}">
             </div>
 
+            {if $tlCfg->authentication['oauth_enabled']}
+            <div class="button">
+            <a style="text-decoration: none; color:#ffffff;" href="{$gui->oauth}">
+            <img src="{$tlCfg->theme_dir}images/google.png" style="height: 42px; vertical-align:middle;">
```
Comment (fmancardi) [inline gui/templates/login-model-marcobiedermann.tpl]: it's not the standard, images has to be added on tlSmarty class

File: gui/templates/login-model-marcobiedermann.tpl [hunk 2]
```diff
@@ -63,6 +63,14 @@
...
+            <span style="padding: 10px;">Sign in with Google</span></a>
```
Comment (fmancardi) [inline gui/templates/login-model-marcobiedermann.tpl]: it's not standard all strings has to be defined on localization files

File: lib/functions/oauth_api.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ * @since 1.9.17
+ *
+ */
+  
+// Create correct link for oauth
+function oauth_link($authCfg)
+{
+  $oauth_url = $authCfg['oauth_url'] . '/auth';
+  $oauth_params = array(
+    'redirect_uri'  => 'http://' . $_SERVER[HTTP_HOST]. '/login.php?oauth=true',
```
Comment (fmancardi) [inline lib/functions/oauth_api.php]: Bad because what will happen if you set https?

File: lib/functions/oauth_api.php [hunk 4]
```diff
@@ -16,17 +16,38 @@
 // Create correct link for oauth
 function oauth_link($authCfg)
 {
+  $promt = 'none';
+  if ($tlCfg->authentication['oauth_force_single'])
```
Comment (fmancardi) [inline lib/functions/oauth_api.php]: our standard is always use {}

**Comment by Kanaduchi** [comment]
Updated PR after comments

**Pull request closed without merging**